### PR TITLE
[Charmhub] - Typed error codes

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following attempts to handle some of the basic error codes from the
charmhub API. This isn't a tautological list but should give us enough
to work with.

This should smarten up the handling of charmhub as a developer quite
considerably.

## QA steps

Test pass.

### CODE REVIEW

File: charmhub/errors.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package charmhub
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/juju/charmhub/transport"
```
Comment (hmlanigan) [inline charmhub/errors.go:8]: nit: 2 stanzas

File: charmhub/errors.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	if list == nil || len(list) == 0 {
+		return nil
+	}
+
+	if errs, _ := APIErrors(list); errs != nil {
+		switch errs[0].Code {
+		case transport.ErrorCodeNotFound, transport.ErrorCodeNameNotFound:
+			return errors.NotFoundf(errs[0].Message)
+		case transport.ErrorCodeAPIError:
+			return errors.NotValidf(errs[0].Message)
```
Comment (hmlanigan) [inline charmhub/errors.go:57]: Due to the really long message potential on ErrorCodeAPIError, can we log the error on the controller and return something different in the NotValidf.

https://pastebin.ubuntu.com/p/Kk7RsSqp2v/

The comment on ErrorCodeBadArgument, applies here too

File: charmhub/errors.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+// Handle some of the basic error messages.
+func handleBasicAPIErrors(list transport.APIErrors, logger Logger) error {
+	if list == nil || len(list) == 0 {
+		return nil
+	}
+
+	if errs, _ := APIErrors(list); errs != nil {
+		switch errs[0].Code {
+		case transport.ErrorCodeNotFound, transport.ErrorCodeNameNotFound:
```
Comment (hmlanigan) [inline charmhub/errors.go:54]: add ErrorCodeResourceNotFound:
{"error-list":[{"code":"resource-not-found","message":"No charm or bundle with name 'testme'."}]}

File: charmhub/errors.go [hunk 4]
```diff
@@ -50,19 +51,31 @@ func handleBasicAPIErrors(list transport.APIErrors, logger Logger) error {
...
 
 	if errs, _ := APIErrors(list); errs != nil {
+		masked := true
+		defer func() {
+			// Only log out the error if we're masking the original error, that
+			// way you can at least find the issue in `debug-log`.
+			// We do this because the original error message can be huge and
+			// verbose, like a java stack trace!
+			if masked {
+				logger.Errorf("charmhub API error %s:%s", errs[0].Code, errs[0].Message)
```
Comment (hmlanigan) [inline charmhub/errors.go:61]: 👍

**Review by hmlanigan — APPROVED** [review]
good updates.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**